File: src/main/java/hiiragi283/api/recipe/CrucibleRecipe.java
package hiiragi283.api.recipe;

import hiiragi283.api.material.MaterialStack;
import hiiragi283.integration.jei.ingredients.HiiragiIngredientTypes;
import mezz.jei.api.ingredients.IIngredients;
import mezz.jei.api.ingredients.VanillaTypes;
import net.minecraft.item.ItemStack;
import org.jetbrains.annotations.NotNull;

public class CrucibleRecipe extends HiiragiRecipe<CrucibleRecipe> {

    private final MaterialStack input;
    private final ItemStack cast;
    private final ItemStack output;

    public CrucibleRecipe(MaterialStack input, ItemStack cast, ItemStack output) {
        this.input = input;
        this.cast = cast;
        this.output = output;
    }

    public MaterialStack getInput() {
        return this.input;
    }

    public ItemStack getCast() {
        return this.cast.copy();
    }

    public ItemStack getOutput() {
        return this.output.copy();
    }

    //    IRecipeWrapper    //


    @Override
    public void getIngredients(@NotNull IIngredients iIngredients) {
        iIngredients.setInput(HiiragiIngredientTypes.MATERIAL, getInput());
        iIngredients.setInput(VanillaTypes.ITEM, getCast());
        iIngredients.setOutput(VanillaTypes.ITEM, getOutput());
    }

}

File: src/main/kotlin/hiiragi283/api/recipe/CrucibleRecipe.kt
package hiiragi283.api.recipe

import hiiragi283.api.material.HiiragiMaterial
import hiiragi283.api.part.HiiragiPart
import hiiragi283.api.part.PartRegistry
import hiiragi283.integration.jei.ingredients.HiiragiIngredientTypes
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.ingredients.VanillaTypes
import net.minecraft.item.ItemStack
import net.minecraftforge.fluids.FluidRegistry
import net.minecraftforge.fluids.FluidStack

class CrucibleRecipe(
    val input: HiiragiMaterial,
    val tempMin: Int,
    private val fluid: String
) : HiiragiRecipe<CrucibleRecipe>() {

    companion object {
        @JvmField
        val EMPTY = CrucibleRecipe(HiiragiMaterial.EMPTY, -1, "water")
    }

    fun isEmpty() = this == EMPTY

    fun getOutput(amount: Int = 1000): FluidStack = FluidStack(FluidRegistry.getFluid(fluid), amount)

    fun getOutput(stack: ItemStack): FluidStack {
        return getOutput((PartRegistry.getParts(stack)
            .firstOrNull { it.material == input }
            ?: HiiragiPart.EMPTY).shape.scale)
    }

    fun matches(stack: ItemStack): Boolean {
        if (stack.isEmpty) return false
        return input in PartRegistry.getParts(stack).map { it.material }
    }

    //    IRecipeWrapper    //

    override fun getIngredients(p0: IIngredients) {
        p0.setInput(HiiragiIngredientTypes.MATERIAL, input)
        p0.setOutput(VanillaTypes.FLUID, getOutput())
    }

}